### Initial Data Exploration and Preparation

In this phase, we aim to familiarize ourselves with the dataset's structure and content. We will also begin the data cleaning process to ensure the dataset is properly formatted and ready for analysis.


In [1]:
 # Importing necessary Libraries
import pandas as pd
import numpy as np

In [5]:
 # Loading the CSV File
DATA = pd.read_csv( 'HR_Employee_Attrition.csv' )

In [12]:
 # Getting the top and bottom rows
display( DATA.head() )
display( DATA.tail() )

 # Getting the number of rows and columns
print( 'Number of Rows : ' ,(DATA.shape)[0] )
print( 'Number of Columns : ' ,(DATA.shape)[1] )

 # Getting the Columns
for i in ( DATA.columns.to_list() ):
    print(i)

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

Age Attrition     BusinessTravel  DailyRate              Department  \
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9                  6   
1469               4               4                  3   

      YearsSinceLastPromotion  YearsWithCurrManager  
1465                        0                     3  
1466                        1                     7  
1467                        0                     3  
1468                        0                     8  
1469                        1                     2  

[5 rows x 35 columns]

Number of Rows :  1470
Number of Columns :  35
Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EmployeeCount
EmployeeNumber
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
Over18
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StandardHours
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


From the output we can see that the dataframe contains
- 1470 rows 
- 35 Columns 

We will get indepth wo thw columns and understand what the columns mean and what type of data they hold.

In [ ]:
 # Getting the datatypes
display( DATA.dtypes.to_frame() )

0
Age                        int64
Attrition                 object
BusinessTravel            object
DailyRate                  int64
Department                object
DistanceFromHome           int64
Education                  int64
EducationField            object
EmployeeCount              int64
EmployeeNumber             int64
EnvironmentSatisfaction    int64
Gender                    object
HourlyRate                 int64
JobInvolvement             int64
JobLevel                   int64
JobRole                   object
JobSatisfaction            int64
MaritalStatus             object
MonthlyIncome              int64
MonthlyRate                int64
NumCompaniesWorked         int64
Over18                    object
OverTime                  object
PercentSalaryHike          int64
PerformanceRating          int64
RelationshipSatisfaction   int64
StandardHours              int64
StockOptionLevel           int64
TotalWorkingYears          int64
TrainingTimesLastYear      int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsInCurrentRole         int64
YearsSinceLastPromotion    int64
YearsWithCurrManager       int64

In [33]:
for i in DATA.columns.to_list() :
    print( i , ' = ' , ( DATA[i].unique() )[0:5] , ' - ', DATA[i].nunique() )  

Age  =  [41 49 37 33 27]  -  43
Attrition  =  ['Yes' 'No']  -  2
BusinessTravel  =  ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']  -  3
DailyRate  =  [1102  279 1373 1392  591]  -  886
Department  =  ['Sales' 'Research & Development' 'Human Resources']  -  3
DistanceFromHome  =  [ 1  8  2  3 24]  -  29
Education  =  [2 1 4 3 5]  -  5
EducationField  =  ['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree']  -  6
EmployeeCount  =  [1]  -  1
EmployeeNumber  =  [1 2 4 5 7]  -  1470
EnvironmentSatisfaction  =  [2 3 4 1]  -  4
Gender  =  ['Female' 'Male']  -  2
HourlyRate  =  [94 61 92 56 40]  -  71
JobInvolvement  =  [3 2 4 1]  -  4
JobLevel  =  [2 1 3 4 5]  -  5
JobRole  =  ['Sales Executive' 'Research Scientist' 'Laboratory Technician'
 'Manufacturing Director' 'Healthcare Representative']  -  9
JobSatisfaction  =  [4 2 3 1]  -  4
MaritalStatus  =  ['Single' 'Married' 'Divorced']  -  3
MonthlyIncome  =  [5993 5130 2090 2909 3468]  -  1349
MonthlyRate  =  [19479 24907  23

From the output, we observe that the dataset primarily consists of two data types: `int64` and `object`. However, some columns that are inherently categorical in nature—such as *Education* and *JobLevel*—are currently represented as `int64`.

Upon further examination:

1. **Categorical Columns**:
   The following columns contain categorical data and are currently represented as objects or integers:

   * `Attrition`, `BusinessTravel`, `Department`, `EducationField`, `Gender`, `JobRole`, `MaritalStatus`, `Over18`, and `OverTime`

2. **Numerical Continuous Columns**:
   These columns represent continuous numerical values:

   * `Age`, `DailyRate`, `DistanceFromHome`, `EmployeeNumber`, `HourlyRate`, `MonthlyRate`, `MonthlyIncome`, `NumCompaniesWorked`, `PercentSalaryHike`, `TotalWorkingYears`, `TrainingTimesLastYear`, `YearsAtCompany`, `YearsInCurrentRole`, `YearsSinceLastPromotion`, and `YearsWithCurrManager`

3. **Numerical Categorical Columns**:
   These are ordinal variables typically measured on a scale of 1 to 5, where 1 generally represents the lowest level of satisfaction, involvement, or performance, and 5 indicates the highest level

   * `Education`, `EnvironmentSatisfaction`, `JobInvolvement`, `JobLevel`, `JobSatisfaction`, `PerformanceRating`, `StockOptionLevel`, and `WorkLifeBalance`

**NEXT**

We will proceed to clean the dataset and convert columns to their appropriate data types to ensure proper analysis and modeling.


In [40]:
 # Getting null values
display( DATA.isnull().sum() )

 # Getting the Duplicated 
print( 'Number of Duplicated Values : ' , DATA.duplicated().sum() )

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

Number of Duplicated Values :  0


From the output, we confirm that the dataset contains no missing (null) values and no duplicate records. This indicates that the data is clean and ready for further analysis.


We will now proceed with Exploratory Data Analysis (EDA) to uncover patterns, trends, and insights within the dataset.